### PR TITLE
Use interface and runtime

### PR DESCRIPTION
**Pull request by jpihl**
To compile use
./waf configure --kodo-use-checkout=make-ready-for-kodo-c

### CODE REVIEW

**Comment by jpihl** [comment]
@petya2164 do you have any comments about this?

File: src/kodoc/kodoc.h [hunk 1]
```diff
@@ -70,8 +68,8 @@ typedef enum
     kodo_sliding_window,
     kodo_sparse_full_vector,
     kodo_seed,
-    kodo_sparse_seed,
```
Comment (petya2164) [inline src/kodoc/kodoc.h]: @jpihl What happened with the sparse_seed codec here? We really need that one ;)

File: src/kodoc/kodoc.h [hunk 2]
```diff
@@ -296,8 +272,16 @@ void kodo_set_symbols(kodo_coder_t encoder, uint8_t* data, uint32_t size);
 /// @param data The buffer containing the data to be encoded
 /// @param size The size of the symbol buffer
 KODOC_API
-void kodo_set_symbol(kodo_coder_t encoder, uint32_t index,
-                     uint8_t* data, uint32_t size);
+void kodo_set_const_symbol(kodo_coder_t encoder, uint32_t index, uint8_t* data,
+    uint32_t size);
+
+KODOC_API
```
Comment (petya2164) [inline src/kodoc/kodoc.h]: We should probably explain the difference between the const and mutable variants here.

**Comment by petya2164** [comment]
In general, I am happy with the simplifications. A lot of complexity has been shifted to kodo itself.
I have a strong aversion against these insane nested namespaces, e.g. a cast to `kodo::api::final_interface::config_interface*`. This almost feels like Java! Let's put some using statements in kodoc.cpp!
Another concern I have is about the code bloat and compilation time. We should check the final size of the shared lib and the total compilation time + RAM usage.
We also have a throughput benchmark here that should be checked before this can be merged. I'm especially concerned for the "Setup time" here: https://github.com/steinwurf/kodo-c/blob/use-interface-and-runtime/benchmark/kodoc_throughput/kodoc_throughput.cpp#L158

File: src/kodoc/create_factory_wrapper.hpp [hunk 3]
```diff
@@ -9,74 +9,83 @@
...
-#include <meta/typelist.hpp>
+template<template<class,class,class...> class Coder, class... Interfaces>
+kodo_factory_t
+create_encoder_factory(int32_t finite_field, uint32_t max_symbols,
+                       uint32_t max_symbol_size, int32_t trace_mode,
+                       int32_t storage_mode)
+{
+    using namespace kodo;
+
+    std::map<int32_t, std::string> finite_field_map = {
```
Comment (petya2164) [inline src/kodoc/create_factory_wrapper.hpp]: Let's put that { to the next line!
I guess this example can be added to our style guide, if there is some ambiguity here.

File: src/kodoc/kodoc.h (hunk 1 above)
Comment (jpihl) [inline src/kodoc/kodoc.h]: we only have this in a shallow version so it didn't really fit the new design. I mostly just removed it to make something able to compile ;) I'll just add it again - I was not really sure if it was used or not.

File: src/kodoc/kodoc.h (hunk 2 above)
Comment (jpihl) [inline src/kodoc/kodoc.h]: sure

**Pull request closed without merging**